File: golang/18-big-bang-algorithm/fraud_detector.go
package kata

import "time"

type Tx struct {
	Amount      float64
	Timestamp   int64
	History     []Tx
	Merchant    string
	Country     string
	CardCountry string
}
type Risk struct {
	Score  int
	Level  int
	Rating string
}

type FraudDetector struct{}

func NewFraudDetector() *FraudDetector {
	return &FraudDetector{}
}

func (fd *FraudDetector) detect(tx Tx) Risk {
	s := 0
	v := 0
	m := 0
	d := time.UnixMilli(tx.Timestamp).UTC().Hour()
	for _, h := range tx.History {
		if h.Amount > tx.Amount*2 {
			v++
		}
		if fd.abs(tx.Timestamp-h.Timestamp) < 3600000 {
			s++
		}
	}
	if tx.Amount > 500 && d >= 0 && d < 6 {
		m += 30
	}
	if tx.Amount > 1000 {
		m += 20
	}
	if tx.Merchant == "gambling" || tx.Merchant == "crypto" {
		m += 25
	}
	if tx.Country != tx.CardCountry {
		m += 15
	}
	if s > 3 {
		m += 20
	}
	if v > 2 {
		m += 15
	}
	level := 1
	if m >= 80 {
		level = 5
	} else if m >= 60 {
		level = 4
	} else if m >= 40 {
		level = 3
	} else if m >= 20 {
		level = 2
	}
	rating := []string{"", "low", "medium", "elevated", "high", "critical"}[level]
	return Risk{m, level, rating}
}
func (fd *FraudDetector) abs(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}

File: ruby/18-big-bang-algorithm/src/fraud_detector.rb
# frozen_string_literal: true

class FraudDetector
  def detect(tx)
    s = 0
    v = 0
    m = 0
    d = Time.at(tx[:timestamp] / 1000).utc.hour
    tx[:history].each do |h|
      v += 1 if h[:amount] > tx[:amount] * 2
      s += 1 if (tx[:timestamp] - h[:timestamp]).abs < 3_600_000
    end
    m += 30 if tx[:amount] > 500 && d >= 0 && d < 6
    m += 20 if tx[:amount] > 1000
    m += 25 if %w[gambling crypto].include?(tx[:merchant])
    m += 15 if tx[:country] != tx[:card_country]
    m += 20 if s > 3
    m += 15 if v > 2
    level = if m < 20
              1
            elsif m < 40
              2
            elsif m < 60
              3
            else
              m < 80 ? 4 : 5
            end
    rating = { 1 => 'low', 2 => 'medium', 3 => 'elevated', 4 => 'high', 5 => 'critical' }[level]
    { score: m, level: level, rating: rating }
  end
end
